tribits/common_tpls/FindTPLHDF5.cmake

```diff
@@ -2,78 +2,97 @@
 # See associated tribits/Copyright.txt file for copyright and license! #
 ########################################################################
 
-#
-# First, set up the variables for the (backward-compatible) TriBITS way of
-# finding HDF5.  These are used in case find_package(HDF5 ...) is not called
-# or does not find HDF5.  Also, these variables need to be non-null in order
-# to trigger the right behavior in the function
-# tribits_tpl_find_include_dirs_and_libraries().
-#
-
-set(REQUIRED_HEADERS hdf5.h)
-set(REQUIRED_LIBS_NAMES hdf5)
-
-if (HDF5_REQUIRE_FORTRAN)
-  set(REQUIRED_LIBS_NAMES ${REQUIRED_LIBS_NAMES} hdf5_fortran)
-endif()
+if (Netcdf_ALLOW_MODERN)
 
-if (TPL_ENABLE_MPI)
-  set(REQUIRED_LIBS_NAMES ${REQUIRED_LIBS_NAMES} z)
-endif()
+  #set(minimum_modern_HDF5_version 4.7.4)
+  print_var(Netcdf_ALLOW_MODERN)
+  message("-- Using find_package(HDF5 ${minimum_modern_HDF5_version} CONFIG) ...")
+  find_package(HDF5  ${minimum_modern_HDF5_version}  CONFIG)
+  if (HDF5_FOUND)
+    message("-- Found HDF5_CONFIG=${HDF5_CONFIG}")
+    message("-- Generating Netcdf::all_libs and NetcdfConfig.cmake")
+    tribits_extpkg_create_imported_all_libs_target_and_config_file(
+      HDF5
+      INNER_FIND_PACKAGE_NAME  HDF5
+      IMPORTED_TARGETS_FOR_ALL_LIBS   HDF5::HDF5)
+  endif()
 
-if (TPL_ENABLE_Netcdf)
-  set(REQUIRED_LIBS_NAMES ${REQUIRED_LIBS_NAMES} hdf5_hl)
 endif()
 
-#
-# Second, search for HDF5 components (if allowed) using the standard
-# find_package(HDF5 ...).
-#
-tribits_tpl_allow_pre_find_package(HDF5  HDF5_ALLOW_PREFIND)
-if (HDF5_ALLOW_PREFIND)
+if (NOT TARGET HDF5::all_libs)
 
-  message("-- Using find_package(HDF5 ...) ...")
+  # First, set up the variables for the (backward-compatible) TriBITS way of
+  # finding HDF5.  These are used in case find_package(HDF5 ...) is not called
+  # or does not find HDF5.  Also, these variables need to be non-null in order
+  # to trigger the right behavior in the function
+  # tribits_tpl_find_include_dirs_and_libraries().
+
+  set(REQUIRED_HEADERS hdf5.h)
+  set(REQUIRED_LIBS_NAMES hdf5)
 
-  set(HDF5_COMPONENTS C)
   if (HDF5_REQUIRE_FORTRAN)
-    list(APPEND HDF5_COMPONENTS Fortran)
+    set(REQUIRED_LIBS_NAMES ${REQUIRED_LIBS_NAMES} hdf5_fortran)
   endif()
 
   if (TPL_ENABLE_MPI)
-    set(HDF5_PREFER_PARALLEL TRUE)
+    set(REQUIRED_LIBS_NAMES ${REQUIRED_LIBS_NAMES} z)
   endif()
 
-  find_package(HDF5 COMPONENTS ${HDF5_COMPONENTS})
-
-  # Make sure that HDF5 is parallel.
-  if (TPL_ENABLE_MPI AND NOT HDF5_IS_PARALLEL)
-    message(FATAL_ERROR "Trilinos is configured for MPI, HDF5 is not.
-    Did CMake find the correct libraries?
-    Try setting HDF5_INCLUDE_DIRS and/or HDF5_LIBRARY_DIRS explicitly.
-    ")
+  if (TPL_ENABLE_Netcdf)
+    set(REQUIRED_LIBS_NAMES ${REQUIRED_LIBS_NAMES} hdf5_hl)
```

Comment:
You could use `list(APPEND)` here.

Reply:
Good point. This was old code so we were just trying to avoid breaking anything.  Also, hopefully, this specialized code will disappear at some point.

Reply:
Because this file is so hard to fully test in all of the use cases, I am not going to change this code at this point. (In the future, it should all be stripped out.)